RFC: "Data flow" models

lamby opened the pull request:
re. #33 

(Not to be merged

lamby committed:
Implement patient postcode.

lamby committed:
Implement patient address lines.

lamby committed:
Implement provider declaration.

lamby committed:
Implement schedule query field.

lamby committed:
Implement three various dates.

lamby committed:
Add a bit of spacing

lamby committed:
Split out known types.

lamby committed:
Also support top-level datetime.date

lamby committed:
Implement types of claims.

lamby committed:
Add a `strbool` utility.

lamby committed:
fp17/utils.py: flake8 file

lamby committed:
Use our new strbool utility for BCDS1 messages.

lamby committed:
Implement the treatment methods.

lamby committed:
Implement the patient charges.

lamby committed:
Be explicit about "required" attribute.

lamby committed:
Implement exception/remission.

lamby committed:
Move exemption/remission code to a number and parse/generate accordingly.

lamby committed:
Drop unused imports.

lamby committed:
Rework how we flatten our input to support nested lists, etc.

lamby committed:
Parse & generate treatments.

lamby committed:
Implement dental chart as a list.

lamby committed:
Add missing whitespace.

lamby committed:
Ensure exrmcd is 2 characters in length.

lamby committed:
Collapse Treatment and Patient into the BCDS1 module.

lamby committed:
Let treatment_arrangements be optional; all of them and individuals.

lamby committed:
Permit exception remission to be optional too.

lamby committed:
Initial stab at the first testcase.

lamby committed:
Create <trtarr/> element lazily as everything defaults to false.

lamby committed:
These should all be upper case.

lamby committed:
Postcode parts should be uppercase too.

lamby committed:
Resubmission count should be 0-prefixed.

lamby committed:
loc should also be 0-prefixed.

lamby committed:
Calculate "noseg" correctly.

lamby committed:
Actually output/print the XML

lamby committed:
Fixup some whitespace

lamby committed:
Check noseg first.

lamby committed:
Move test to a @pytest.fixture

lamby committed:
Rename BCDS1Message -> BCDS1

lamby committed:
Make it more obvious this takes an XSD schema.

lamby committed:
Scaffolding of Envelope message.

lamby committed:
We always need a schvn.

lamby committed:
We always need a synv too.

lamby committed:
Implement the <ic/> envelope origin field.

lamby committed:
Drop these BCDS1-specific part of the Envelope tests.

lamby committed:
Implement 'destination'

lamby committed:
Implement date of release.

lamby committed:
Implement time of release.

lamby committed:
Merge date and time in our interface into a datetime.

lamby committed:
Implement the serial number.

lamby committed:
Implement swver <ic/> element

lamby committed:
Actually, this *should* be required.

lamby committed:
Implement software package name.

lamby committed:
More whitespace OCD...

lamby committed:
Reorder some elements of the schema.

lamby committed:
datrel should be YYMMDD, not YYYMMDD

lamby committed:
Test that we do some validation at least.

lamby committed:
Test our utils.

lamby committed:
Test we support zero

lamby committed:
Ignore elements starting with _ recursively.

lamby committed:
Support wrapping things!

lamby committed:
Implement approval number.

lamby committed:
Implement revision number.

lamby committed:
Implement transmission category.

lamby committed:
Implement test indicator.

lamby committed:
Explicitly specify required: False.

lamby committed:
Implement interchange control count.

lamby committed:
Refactor loop.

lamby committed:
Enforce that destination is a string.

lamby committed:
pmsno is not required.

lamby committed:
Generate test case with <ic/> wrapper.

lamby committed:
Move FIXME to a more prominent location.

lamby committed:
Set sane defaults for Treatment.{instance_count,teeth}

lamby committed:
This is, of course, mad.

lamby committed:
Validate ourselves before generating XML.

lamby committed:
Add reference to processing rules document.

lamby committed:
Yes, this is utterly mad.

lamby committed:
Permit setting attributes via kwargs in __init__

lamby committed:
Save some whitespace after the previous commit.

lamby committed:
Instance count is optional now.

lamby committed:
Case 1!

lamby committed:
Try using some "constants".

lamby committed:
Correct name of document.

lamby committed:
These fields must be uppercase; but validate them that way instead of munging them automatically.

lamby committed:
Add validation for supporting details field.

lamby committed:
Fix test for supporting details field.

lamby committed:
9150, not 9160!

lamby committed:
Add all the treatment categories.

lamby committed:
RECALL_INTERVAL takes a number of months, it is not fixed.

lamby committed:
Case 2.

lamby committed:
Factor out some common things from both testcases.

lamby committed:
Use our common foo.

lamby committed:
Set default for charge amount.

lamby committed:
Currency is already defaulting to GBP

lamby committed:
Case 3.

lamby committed:
Drop comment for now.

lamby committed:
Script to generate all testcases.

lamby committed:
Also generate testcases in Travis.

lamby committed:
Correct "spelling" of exemption.

lamby committed:
Add whitespace

lamby committed:
Add whitespace

lamby committed:
Add explanation.

lamby committed:
Add the number to the ethnic origin.

lamby committed:
Add the original treatments text.

lamby committed:
Drop treatment suffix; we are defining this in a file called treatments.py

lamby committed:
Add some more constants.

lamby committed:
Reflow long comment.

lamby committed:
Initial draft of all testcases. (Incomplete treatments and codes, etc.)

lamby committed:
Drop whitespace.

lamby committed:
Add some more treatment constants.

lamby committed:
Fill in case #4

lamby committed:
Add denture constant classes.

lamby committed:
Fill in cases 5 -> 10.

lamby committed:
Fill in case 13.

lamby committed:
Informally mark cases 18 - 21 as complete.

lamby committed:
Tidy up the RECALL_INTERVAL Treatment.

lamby committed:
Add REFERRAL_FOR_ADVANCED_MANDATORY_SERVICES{,_LEGACY} Treatments

lamby committed:
Add PERMANENT_FILLINGS_AND_SEALANT_RESTORATIONS Treatment.

lamby committed:
Add EXTRACTIONS Treatment.

lamby committed:
Complete testcase 11

lamby committed:
Update RECALL_INTERVAL Treatment callsites to match renamed argument.

lamby committed:
Complete case 12.

lamby committed:
Complete case 14.

lamby committed:
Use EXTRACTION (not EXTRACTIONS) as that's what the documents say.

lamby committed:
Add Sedation Services as a treatment.

lamby committed:
Complete case 15.

lamby committed:
Complete case 16.

lamby committed:
Fill in treatments for case 17.

lamby committed:
Complete case 22.

lamby committed:
Complete case 23.

lamby committed:
Add DOMICILIARY_SERVICES treatment.

lamby committed:
Complete case 24.

lamby committed:
Complete case 25.

lamby committed:
Drop random punctuation.

lamby committed:
This should be a class.

lamby committed:
Add new BRIDGES_FITTED treatment.

lamby committed:
Complete case 26.

lamby committed:
Complete case 28.

lamby committed:
Complete treatments for case 29.

lamby committed:
Complete case 30.

lamby committed:
Add OTHER_TREATMENT treatment.

lamby committed:
Complete treatments for case 31.

lamby committed:
Add new FISSURE_SEALANTS treatment.

lamby committed:
Complete treatments for case 32.

lamby committed:
Move TREATMENT_CATEGORY_BAND_X to a class.

lamby committed:
Add FREE_REPAIR_REPLACEMENT Treatment.

lamby committed:
Complete case 29.

lamby committed:
Complete case 32.

lamby committed:
Add FURTHER_TREATMENT_WITHIN_TWO_MONTHS treatment service.

lamby committed:
Complete case 31.

lamby committed:
Complete services for case 27.

lamby committed:
Add a BEST_PRACTICE_PREVENTION Treatment.

lamby committed:
Complete case 33.

lamby committed:
Complete case 34.

lamby committed:
Complete case 35.

lamby committed:
Complete case 36.

lamby committed:
Complete case 37.

lamby committed:
Complete treatments for case 38.

lamby committed:
Complete treatments for case 40.

lamby committed:
Add treatments for case 43.

lamby committed:
"Band 4" appears to refer to 'urgent'.

lamby committed:
Add some constants for Exemptions and Remissions.

lamby committed:
Actually move these into an `exemptions` module and make them classes.

lamby committed:
Complete "excemptions" for a bunch of cases.

lamby committed:
Correct spelling of "exemption"

lamby committed:
Support generation of a zip of all the testcases.

lamby committed:
Don't ignore .zip files anymore; we are about to not care about them.

lamby committed:
The message reference number can be from 1-6 digits really.

lamby committed:
Location can be 0-prefixed.

lamby committed:
Make supplier_testing a package.

lamby committed:
Add click to requirements.

lamby committed:
Drop common.py utilities for now.

lamby committed:
Update the API of all individual cases to take a message, rather than generate one.

lamby committed:
Update generate script to print supplier testing data in one "file".

lamby committed:
Update .travis.yml for the new script.

lamby committed:
Update tests for 1048eee346fa67d52e221d34c42f847bb434a74d

lamby committed:
teind (and others) should be '1' or '0', not 'true' or 'false'.

lamby committed:
clrn should be 6 characters.

lamby committed:
Include testing data in a single XML Envelope, not a single one.

davidmiller committed:
Notify the notifications channel of a relevant Slack

lamby committed:
Ensure test for message reference number also sets the right length

lamby committed:
Update validation; swname and swversion must be alphanumeric.

lamby committed:
Use "unknown" (not "(unknown)") as the "unknown" version.

davidmiller committed:
Update and rename README.rst to README.md

lamby committed:
pmsno is apparently required, despite what the schema says.

lamby committed:
Add an approval number to the supplier tests.

lamby committed:
Add a real (?) response to test failures as a test fixture.

lamby committed:
Set treatment category to "3" in case #3 to prevent "No significant treatment on an EDI claim" processing error.

lamby committed:
Add a Treatment validation framework.

lamby committed:
Drop whitespace.

lamby committed:
Validate that legacy advanced mandatory services are not used after 01/04/2014

lamby committed:
Update supplier tests to use the correct mandatory services code.

davidmiller committed:
Actually convert the badge to markdown

lamby committed:
More RST -> Markdown conversion (re. #2)

lamby committed:
Correct rendering and URI of Apperta Foundation link. (re. #2)

davidmiller committed:
Note secondary installation of pytest, lift magic pytest incantation from .travis and add introductory notes.

davidmiller committed:
Fixup: We try for spaces in OHC; use emails that exist

davidmiller committed:
Fixup: O H C rather than OHC when expanded

davidmiller committed:
Introductory Sphinx build instructions

davidmiller committed:
Fixup: O H C rather than OHC. Definitely no commas :)

davidmiller committed:
Fixup: Typo

davidmiller committed:
Install pytest-cov extension on Travis

davidmiller committed:
Trigger coveralls on success. (Closes: #9)

davidmiller committed:
Ignore coverage artifacts

lamby committed:
Add a comment regarding why the strange-looking return-and-yield in the same method. (Closes: #7)

lamby committed:
Drop code leftover from trying a few approaches to a `Treatment` validation framework. (Closes: #8)

Thanks, David!

lamby committed:
Add dependencies to setup.py. (Closes: #6)

lamby committed:
Drop errant top-level index.rst.

lamby committed:
Explicitly label items for Sphinx automoduling, including specifying :members:. (Closes: #5)

lamby committed:
Ensure the documentation at least builds.

lamby committed:
Can skip a documentation step with Make's `-C` argument. (Re. #3)

fredkingham committed:
adds the fpt17 requirements to the odonto project requirements

lamby committed:
WIP: "Data flow" models. (Not to be merged)

fredkingham commented on odonto/models.py:
```diff
@@ -187,3 +187,119 @@ class Fp17Declaration(models.EpisodeSubrecord):
+    user = models.ForeignKey(
+        'auth.User',
+        on_delete=models.CASCADE,
+        related_name='',
```
what happens when you put in an empty related_name?

lamby replied:
It fails. (These are super draft models!)

lamby committed:
Drop comment; we'll do this at the PR level, or simply by using common sense..

lamby committed:
Merge branch 'master' into data-flow-models

* master:
  Add some trailing commas.
  Remove trailing whitespaces. (PEP8 W291)
  Put colon-separated compound statement on separate lines. (PEP8 E701)
  Try to make lines fit within --max-line-length characters. (PEP8 E501)
  Add missing 2 blank lines. (PEP8 E302)
  Fix extraneous whitespace around keywords and operators. (PEP8 E221/E225)
  Fix a badly indented lines. (PEP8 E121/E122)
  Reindent all lines, fixing trailing-file newlines. (PEP8 E101)

# Conflicts:
#	odonto/models.py

lamby committed:
Switch SubmissionBCDS1Message -> BCDS1MessageSubmission.

lamby committed:
Reorder for next commit.

lamby committed:
Drop unnecessary quotes.

lamby committed:
Correct spelling error (!).

lamby committed:
Correct foreign key model name.

lamby committed:
Namespace related_name more accurately.

lamby committed:
State belongs to the submission.

lamby committed:
Add a tri-state for the state field.

lamby committed:
We'll probably get a response, so lets save that.

lamby committed:
Add an "updated" field as we'll be updating the state and the response.

Signed-off-by: Chris Lamb <chris@openhealthcare.org.uk>

lamby committed:
Move the state/response to the MCDS1MessageSubmission.

lamby committed:
Add the "updated" field to the BCDS1MessageSubmission.

lamby committed:
Move data-flow related stuff to its own app.

lamby committed:
Drop "User" field.

lamby committed:
Add 'python-fp17/' from commit '254c7ea1d0a1f46be24b35e73807d5e5943fd0d6' via subtree merge.

git-subtree-dir: python-fp17
git-subtree-mainline: 4236fa25a66b4fd2cfeee9cd93ead9fbd3d1461d
git-subtree-split: 254c7ea1d0a1f46be24b35e73807d5e5943fd0d6

lamby committed:
Blindly move the fp17 module to the top-level.

lamby committed:
Drop now-duplicate requirements.txt.

lamby committed:
Drop now-unnecessary distribution-related foo.

lamby committed:
Merge READMEs.

lamby committed:
Add WIP new .travis.yml

Signed-off-by: Chris Lamb <chris@openhealthcare.org.uk>

lamby committed:
Drop old .travis.yml for fp17

lamby committed:
Generate documentation.

lamby committed:
Drop old "docs" dir.

lamby committed:
Exclude the generated Sphinx documentation from the git repo.

lamby committed:
Move these tests, etc. under the right Odonto app.

lamby committed:
Drop nested .gitignore.

lamby committed:
Add some migrations, at least for sanity checking during tests.

lamby committed:
Don't use deprecated TEMPLATE_LOADERS in settings.

lamby committed:
Correct TestCase typo.

lamby committed:
Fix importing under Python 2.x

lamby committed:
Hack around lack of nonlocal in Python 2.x

lamby committed:
Correct nonlocal reference.

lamby committed:
Fix more Python 2.x compatbility issues.

lamby committed:
Correct .xsd location path.

lamby committed:
More Python 2.x compatibility fixes.

lamby committed:
Initial port of tests to Django-style. Well, OpelTestCase.

fredkingham commented:
closing for the time being
